staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go

```diff
@@ -97,6 +111,40 @@ func (a *cachedTokenAuthenticator) AuthenticateToken(ctx context.Context, token
 	return resp, ok, err
 }
 
-func keyFunc(auds []string, token string) string {
-	return fmt.Sprintf("%#v|%v", auds, token)
+// keyFunc generates a string hash based on the inputs.
+// This lowers the memory requirement of the cache and keeps tokens out of memory.
+func keyFunc(randomCachePrefix []byte, auds []string, token string) string {
+	// allocate a single buffer based on the longest string
+	maxLength := len(token)
+	for _, aud := range auds {
+		if length := len(aud); length > maxLength {
+			maxLength = length
+		}
+	}
+	b := make([]byte, 4+maxLength) // include extra bytes to store the length
+
+	h := sha256.New()
+
+	if _, err := h.Write(randomCachePrefix); err != nil {
+		panic(err) // Write() on hash never fails
+	}
+
+	hashLengthPrefixedString(h, b, token)
+	for _, aud := range auds {
```

Comment:
Are auds guaranteed to be in the same order every time?

I think since auds is the last thing to go in to the hash, this code is OK. If something were to be appended, though, I think we'd also want to first write in the number of auds, otherwise it'd be possible to get a collision if the last aud could be the same as that other thing.